modules/clickhouse/clickhouse_test.go

```diff
@@ -0,0 +1,274 @@
+
+	ch "github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/stretchr/testify/assert"
+)
+
+const dbname = "testdb"
+const user = "clickhouse"
+const password = "password"
+
+const createTableQuery = "create table if not exists test_table (id UInt64) engine = MergeTree PRIMARY KEY (id) ORDER BY (id) SETTINGS index_granularity = 8192;"
+const insertQuery = "INSERT INTO test_table (id) VALUES (1);"
+const selectQuery = "SELECT * FROM test_table;"
+
+type Test struct {
+	Id uint64
+}
+
+func TestClickHouseDefaultConfig(t *testing.T) {
+	ctx := context.Background()
+
+	container, err := RunContainer(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Clean up the container after the test is complete
+	t.Cleanup(func() {
+		assert.NoError(t, container.Terminate(ctx))
+	})
+
+	connectionHost, err := container.ConnectionHost(ctx)
+	assert.NoError(t, err)
+
+	conn, err := ch.Open(&ch.Options{
+		Addr: []string{connectionHost},
+		Auth: ch.Auth{
+			Database: container.dbName,
+			Username: container.user,
+			Password: container.password,
+		},
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+	defer conn.Close()
+
+	err = conn.Ping(context.Background())
+	assert.NoError(t, err)
+}
+
+func TestClickHouseIpPort(t *testing.T) {
+	ctx := context.Background()
+
+	// customInitialization {
+	container, err := RunContainer(ctx,
+		WithUsername(user),
+		WithPassword(password),
+		WithDatabase(dbname),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// }
+
+	// Clean up the container after the test is complete
+	t.Cleanup(func() {
+		assert.NoError(t, container.Terminate(ctx))
+	})
+
+	// connectionHost {
+	connectionHost, err := container.ConnectionHost(ctx)
+	assert.NoError(t, err)
+	// }
+
+	conn, err := ch.Open(&ch.Options{
+		Addr: []string{connectionHost},
+		Auth: ch.Auth{
+			Database: dbname,
+			Username: user,
+			Password: password,
+		},
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+	defer conn.Close()
+
+	// perform assertions
+	err = conn.Exec(context.Background(), createTableQuery)
+	assert.NoError(t, err)
+
+	err = conn.Exec(context.Background(), insertQuery)
+	assert.NoError(t, err)
+
+	rows, err := conn.Query(context.Background(), selectQuery)
+	assert.NoError(t, err)
+	assert.NotNil(t, rows)
+
+	var data []Test
+	for rows.Next() {
+		var r Test
+		err := rows.Scan(&r.Id)
+
+		assert.NoError(t, err)
+
+		data = append(data, r)
+	}
+
+	assert.Len(t, data, 1)
+}
+
+func TestClickHouseDSN(t *testing.T) {
+	ctx := context.Background()
+
+	container, err := RunContainer(ctx, WithUsername(user), WithPassword(password), WithDatabase(dbname))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Clean up the container after the test is complete
+	t.Cleanup(func() {
+		assert.NoError(t, container.Terminate(ctx))
+	})
+
+	// connectionString {
+	connectionString, err := container.ConnectionString(ctx, "debug=true")
+	assert.NoError(t, err)
+	// }
+
+	opts, err := ch.ParseDSN(connectionString)
+	assert.NoError(t, err)
+
+	conn, err := ch.Open(opts)
```

Comment:
Wdyt if you define a different test function for the code from here to the end of the current test function? Something to represent that it creates tables, inserts data and retrieves it: `TestClickHouseCRUD`.

At the same time, I see duplication in the query constants and the init-db script. Why not using the `WithInitScripts`. I can imagine you want to the the `conn.Exec` method works, right? I'd call a `DELETE` or `UPDATE`  (I'm not a ClickHouse user at all!! 🙏 ) to check the `Exec` call, and then the `conn.Query` method.

But this is not a blocker, I'd like to know your thoughts on this.

Reply:
About `TestClickHouseCRUD`, I wrote about it in my previous comment.

I tested `WithInitScripts` in `TestClickHouseWithInitScripts` function; and in other tests, I prefer to not use it so I wanted to use Exec like using it in a regular application code. I think testing `Exec` is fine now because I wanted to do similar work as in other test functions. I don't think testing `Exec` with `DELETE` or `UPDATE` makes any difference anyway. 

They are my thoughts, but I would like to know if you think I am wrong in some place.

Reply:
> I prefer to not use it so I wanted to use Exec like using it in a regular application code

This makes total sense, thanks for the explanation :) 

And I agree with you: I prefer duplication in tests for the sake of clarity. Let's keep them as they are.